### ISSUE TITLE
CMake: precice-tools links boost libs even though they are specified PRIVATE in target precice

### ISSUE THREAD

**Issue by astadik** [precice/precice#1597]
Operating system Ubuntu20.04
preCICE Version: 2.5.0

I am compiling preCICE and all dependencies via conan
Boost is at version 1.71.0 all libraries but boost-python and -MPI  are compiled

All targets are properly specified and their dependencies are correctly defined in INTERFACE_LINK_LIBRARIES

When conan reaches compilation step of precice-tools (after successfully compiling target precice) it fails with

```
[ 35%] Linking CXX shared library libprecice.so
[ 35%] Built target precice
Scanning dependencies of target precice-tools
[ 35%] Building CXX object CMakeFiles/precice-tools.dir/src/drivers/main.cpp.o
Scanning dependencies of target testprecice
[ 35%] Linking CXX executable precice-tools
/usr/bin/ld: warning: libboost_regex.so.1.71.0, needed by ~/.conan/data/boost/1.71.0/essteyr/boost/package/abd3ca9581f5ec3d6672fa2ee8818b1f09dbb082/lib/libboost_log_setup.so.1.71.0, not found (try using -rpath or -rpath-link)
/usr/bin/ld: ~/.conan/data/boost/1.71.0/essteyr/boost/package/abd3ca9581f5ec3d6672fa2ee8818b1f09dbb082/lib/libboost_log_setup.so.1.71.0: undefined reference to `boost::re_detail_107100::cpp_regex_traits_implementation<wchar_t>::transform[abi:cxx11](wchar_t const*, wchar_t const*) const'
```

when changing linkage of Boost in precice to PUBLIC everything compiles successfuly (line 278)
<code>target_link_libraries(precice PUBLIC</code>

Since it is labeled as PRIVATE but it actually links symbols from Boost libraries, I would guess its actually in target precice PUBLIC API, or directly used somewhere in target precice-tools code?

**Comment by fsimonis** [comment]
Hi, interesting to see a rising Conan package!

Strange error though.

This shouldn't happen as you are building preCICE as a shared library (which is good).
precice-tools is designed to rely only on the STL and libprecice itself.
All dependencies of preCICE private by design.
This basically means, that the linker is doing all the work.

In your case, libprecice.so links without issues, meaning that the dependencies are all detected correctly by the linker.
precice-tools however fails to link due to the linker failing to find boost regex requested by boost log.
The timing of this error doesn't make sense as it should have occurred when linking libprecice.

I recommend you to do the following:
* Try to build only libprecice
* Then try to build only precice-tools and see if the error persists.
* If so, then try to look for things that _look wrong_ when inspecting libprecice.so with [lddtree](https://github.com/ncopa/lddtree).

You may also want to try and see if this problem persists on develop.

**Comment by astadik** [comment]
Thanks for the fast reply, 

libprecice.so compiles fine, probably since it links Boost::log which has in its INTERFACE_LINK_LIBRARIES Boost::regex (for my version of boost conanfile.py). I checked the linkage with ldd and readelf -d and found that libprecice.so due to usage of RUNPATH doesn't find boost_regex. 
I modified it so cmake tells gcc to write RPATH, then ldd found everything, but it didn't fix the issue. (lddtree still shos not found)

I also patches the CMakeLists.txt in conanfile to also link Boost::regex but that simply got optimized away.
Only thing that worked was setting Boost libs as PUBLIC instead of PRIVATE then everything compiles, probably since INTERFACE_LINK_LIBRARIES get forwarded to other targets in this case.

As a note I also tried with Boost 1.71 from conan center, but there it is even worse as it already fails due to not linking Boost::log, I couldn't find any way to fix this (probably since package info is missing linkage info between boost libs)

I still didn't try with develope, I will try that.
